File: introADT/groups.py
class Set:
    def __init__(self, *ary):
        self.entries = list(ary)

    def __add__(self, enum):
        import copy
        r = copy.copy(self.entries)
        [r.append(e) for e in enum.entries if e not in self.entries]
        return Set(*r)

    def __sub__(self, enum):
        return Set(*[e for e in self.entries if e not in enum.entries])

    def __iter__(self):
        for entry in self.entries:
            yield entry

arias = Set("a", "b", "c", "d")
otro =  Set("b", "e", "f", "g", "h")

print "Todos"
for alumno in arias + otro:
    print alumno

print "Parejas"
for pareja in zip(arias, otro - arias):
    print pareja

File: introADT/groups.rb
class Set

    #crea un accesor y mutador
    attr_accessor :entries

    #este operador (*) sirve para recibir una cantidad variable de argumentos
    def initialize(*ary)
        @entries = ary
    end

    def +(enum)
        r = @entries.dup
        enum.entries.each do |e|
            r << e unless r.include? e
        end
        Set.new(*r)
    end

    def -(enum)  
        Set.new(*@entries.reject{|e| enum.entries.include? e})
    end

    def each
        @entries.each{|e| yield e}
    end

end

arias = Set.new *['a', "b", 'c', "d"]
otro  = Set.new *%w{b e f g h}

puts "Todos:"
(arias + otro).each {|alumno| puts alumno}

puts "Parejas:"
p ((arias.entries.sort).zip((otro - arias).entries.sort))
